File: week_2/day_09_monotonic/variations_que/sumSubarrayMinimums.js
function sumSubarrayMins(arr) {
    const mod = 1e9 + 7;
    const stack = [];
    let res = 0;
    arr.push(0);

    for (let i = 0; i < arr.length; i++) {
        while (stack.length && arr[i] < arr[stack[stack.length - 1]]) {
            const mid = stack.pop();
            const left = mid - (stack.length ? stack[stack.length - 1] : -1);
            const right = i - mid;
            res = (res + arr[mid] * left * right) % mod;
        }
        stack.push(i);
    }
    return res;
}

File: week_2/day_09_monotonic/variations_que/sum_subarray_minimums.py
def sumSubarrayMins(arr):
    mod = 10**9 + 7
    stack = []
    res = 0
    arr.append(0)

    for i in range(len(arr)):
        while stack and arr[i] < arr[stack[-1]]:
            mid = stack.pop()
            left = mid - (stack[-1] if stack else -1)
            right = i - mid
            res += arr[mid] * left * right
        stack.append(i)

    return res % mod


# INPUT
print(sumSubarrayMins([3,1,2,4]))
